From 412ed690357f06a9ddf2195fef7868bf005f9c7c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?P=C3=A1draig=20Brady?= <P@draigBrady.com>
Date: Wed, 5 Jan 2011 11:52:54 +0000
Subject: [PATCH] join: print fields in the correct order for unpaired lines

* src/join.c (prfields): A new function refactored from prjoin(),
to output all but the join field.
(prjoin): Don't swap line1 and line2 when line1 is blank
so that the padding is applied to the right place.

Upstream-commit: d4db0cb1827730ed5536c12c0ebd024283b3a4db
Signed-off-by: Kamil Dudka <kdudka@redhat.com>
---
 src/join.c | 65 +++++++++++++++++++++++++++++++++-----------------------------
 1 file changed, 35 insertions(+), 30 deletions(-)

diff --git a/src/join.c b/src/join.c
index 88cdcd1..5bb0fe2 100644
--- a/src/join.c
+++ b/src/join.c
@@ -754,8 +754,6 @@ prfield (size_t n, struct line const *line)
     fputs (empty_filler, stdout);
 }
 
-/* Print the join of LINE1 and LINE2.  */
-
 #define PUT_TAB_CHAR							\
   do									\
     {									\
@@ -764,10 +762,33 @@ prfield (size_t n, struct line const *line)
     }									\
   while (0)								
 
+/* Output all the fields in line, other than the join field.  */
+
+static void
+prfields (struct line const *line, size_t join_field)
+{
+  size_t i;
+
+  for (i = 0; i < join_field && i < line->nfields; ++i)
+    {
+      PUT_TAB_CHAR;
+      prfield (i, line);
+    }
+  for (i = join_field + 1; i < line->nfields; ++i)
+    {
+      PUT_TAB_CHAR;
+      prfield (i, line);
+    }
+}
+
+/* Print the join of LINE1 and LINE2.  */
+
 static void
 prjoin (struct line const *line1, struct line const *line2)
 {
   const struct outlist *outlist;
+  size_t field;
+  struct line const *line;
 
   outlist = outlist_head.next;
   if (outlist)
@@ -777,9 +798,6 @@ prjoin (struct line const *line1, struct line const *line2)
       o = outlist;
       while (1)
         {
-          size_t field;
-          struct line const *line;
-
           if (o->file == 0)
             {
               if (line1 == &uni_blank)
@@ -808,37 +826,24 @@ prjoin (struct line const *line1, struct line const *line2)
     }
   else
     {
-      size_t i;
-
       if (line1 == &uni_blank)
         {
-          struct line const *t;
-          t = line1;
-          line1 = line2;
-          line2 = t;
-        }
-      prfield (join_field_1, line1);
-      for (i = 0; i < join_field_1 && i < line1->nfields; ++i)
-        {
-          PUT_TAB_CHAR;
-          prfield (i, line1);
+          line = line2;
+          field = join_field_2;
         }
-      for (i = join_field_1 + 1; i < line1->nfields; ++i)
+      else
         {
-          PUT_TAB_CHAR;
-          prfield (i, line1);
+          line = line1;
+          field = join_field_1;
         }
 
-      for (i = 0; i < join_field_2 && i < line2->nfields; ++i)
-        {
-          PUT_TAB_CHAR;
-          prfield (i, line2);
-        }
-      for (i = join_field_2 + 1; i < line2->nfields; ++i)
-        {
-          PUT_TAB_CHAR;
-          prfield (i, line2);
-        }
+      /* Output the join field.  */
+      prfield (field, line);
+
+      /* Output other fields.  */
+      prfields (line1, join_field_1);
+      prfields (line2, join_field_2);
+
       putchar ('\n');
     }
 }
-- 
2.9.3

